solve issue #199, improve issue #200 in Winterrupts

joverbee opened the pull request:
#199 corrected an error in the behaviour of attachinterrupt when changing the trigger mode of an interrupt. I now first erase the relevant trigger mode bits, before OR-writing a new setting. (this erasing was missing, causing unpredictable behaviour when calling attachinterrupt more than once for the same pine and with different trigger modes.

#200 I noticed that the EIC_Handler had to loop over all possible external interrupts and check if an interrupt occurred even if we didn't enable those interrupt. This resulted in up to 6us wait times before e.g. EXTINT14 was serviced. I rewrote the handler to keep a list of all attached interrupts (and a handle to their ISRs) with a first come first serve policy, meaning that the first attached interrupt is treated first. This improves the wait time for a single interrupt with edge triggering to slightly less than 2us. This is still long, but already a clear improvement. My guess is that the jump to an ISR with 32 bit address that is stored in RAM is now the main culprit of the remaining time, but this would probably need some inline assembly or a better grasp of how the compiler turns this code into seemingly inefficient assembly code...but hey a factor of 3 is already something. Remember that speed might be an important reason why people decided to buy the Zero (or variants) in the first place.

facchinm commented:
The rationale and the patches look good, I'd only make some slight adjustments.
* remove the debug code `getISRcallback()` and `getISRlist()`and relevant variables or enclose them in a `#if 0` or similar
* separate the patches so every patch addresses a single issue

joverbee commented:
I am quite inexperienced with using Github for collaborative projects so excuse if I am doing this wrong. I reverted the changes and made them in 2 separate commit steps, one solving issue #199 and the other bigger change for improving #200. I removed the debug code which also means Winterrups.h is not changed.

cmaglie commented on cores/arduino/WInterrupts.c:
```diff
@@ -76,41 +81,64 @@ void attachInterrupt(uint32_t pin, voidFuncPtr callback, uint32_t mode)
+    //check if we already have this interrupt
+    int id=-1;
+    for (uint32_t i=0; i<nints; i++){
+      if (ISRlist[i]==in) id=in;
```
The condition here should be `id = i` I guess?

cmaglie commented on cores/arduino/WInterrupts.c:
```diff
@@ -132,25 +160,37 @@ void detachInterrupt(uint32_t pin)
+  //and take it out of the ISR list
+  int id=-1; //find its location first
+  for (uint32_t i=0; i<nints; i++){
+      if (ISRlist[i]==in) id=in;
```
Same here it too should be `id=i`

cmaglie commented on cores/arduino/WInterrupts.c:
```diff
@@ -132,25 +160,37 @@ void detachInterrupt(uint32_t pin)
+  }
+  //and remove the top item
+  ISRlist[nints]=0;
+  ISRcallback[nints]=NULL;
```
Is erasing the list element really needed? IMHO just decreasing `nints` should be enough, new elements added later will overwrite the old value.

cmaglie commented on cores/arduino/WInterrupts.c:
```diff
@@ -132,25 +160,37 @@ void detachInterrupt(uint32_t pin)
+    if ((EIC->INTFLAG.reg & 1<<ISRlist[i]) != 0)
+    {  
+      ISRcallback[i]();// Call the callback function if assigned
+      EIC->INTFLAG.reg = 1<<ISRlist[i]; // Clear the interrupt
```
I'm wondering if we can gain something if we store `1 << ISRlist[i]` instead of `ISRlist[i]`.

joverbee commented:
great! glad somebody is fixing this and apologies for my cumbersome interaction with the github process

regards,
Jo

On 5 Sep 2017, at 14:07, Cristian Maglie <notifications@github.com<mailto:notifications@github.com>> wrote:


@cmaglie commented on this pull request.

Hi @joverbee<https://github.com/joverbee>,
I've added some inline comments, I'm actually working on my local copy, so I'll open another PR with a clean tree that includes this fix.

________________________________

In cores/arduino/WInterrupts.c<https://github.com/arduino/ArduinoCore-samd/pull/201#discussion_r136967788>:

> @@ -76,41 +81,64 @@ void attachInterrupt(uint32_t pin, voidFuncPtr callback, uint32_t mode)
   // Assign pin to EIC
   pinPeripheral(pin, PIO_EXTINT);

-  // Assign callback to interrupt
-  callbacksInt[in] = callback;
+  if (callback){ //only store when there is really an ISR to call
+      //this allow for calling attachInterrupt(pin, NULL, mode), we set up all needed register
+      //but won't service the interrupt, this way we also don't need to check it inside the ISR
+
+    // Store interrupts to service in order of when they were attached
+    // to allow for first come first serve handler
+    uint32_t current=0;
+    //check if we already have this interrupt
+    int id=-1;
+    for (uint32_t i=0; i<nints; i++){
+      if (ISRlist[i]==in) id=in;


The condition here should be id = i I guess?

________________________________

In cores/arduino/WInterrupts.c<https://github.com/arduino/ArduinoCore-samd/pull/201#discussion_r136968085>:

> @@ -132,25 +160,37 @@ void detachInterrupt(uint32_t pin)

   // Disable wakeup capability on pin during sleep
   EIC->WAKEUP.reg &= ~(1 << in);
+
+  //and take it out of the ISR list
+  int id=-1; //find its location first
+  for (uint32_t i=0; i<nints; i++){
+      if (ISRlist[i]==in) id=in;


Same here it too should be id=i

________________________________

In cores/arduino/WInterrupts.c<https://github.com/arduino/ArduinoCore-samd/pull/201#discussion_r136968410>:

> @@ -132,25 +160,37 @@ void detachInterrupt(uint32_t pin)

   // Disable wakeup capability on pin during sleep
   EIC->WAKEUP.reg &= ~(1 << in);
+
+  //and take it out of the ISR list
+  int id=-1; //find its location first
+  for (uint32_t i=0; i<nints; i++){
+      if (ISRlist[i]==in) id=in;
+  }
+  if (id==-1) return; //apparently we didn't have it
+  for (uint32_t i=id; i<nints-1; i++){ //and shift the ones above one down
+      ISRlist[i]=ISRlist[i+1];
+      ISRcallback[i]=ISRcallback[i+1];
+  }
+  //and remove the top item
+  ISRlist[nints]=0;
+  ISRcallback[nints]=NULL;


Is erasing the list element really needed? IMHO just decreasing nints should be enough, new elements added later will overwrite the old value.

________________________________

In cores/arduino/WInterrupts.c<https://github.com/arduino/ArduinoCore-samd/pull/201#discussion_r136968985>:

> -    {
-      // Call the callback function if assigned
-      if (callbacksInt[i]) {
-        callbacksInt[i]();
-      }
-
-      // Clear the interrupt
-      EIC->INTFLAG.reg = 1 << i;
+  //calling the routine directly from -here- takes about 1us
+  //depending on where you are in the list it will take longer
+  for (uint32_t i=0; i<nints; i++) // Loop over all enabled interrupts in the list
+  {
+    if ((EIC->INTFLAG.reg & 1<<ISRlist[i]) != 0)
+    {
+      ISRcallback[i]();// Call the callback function if assigned
+      EIC->INTFLAG.reg = 1<<ISRlist[i]; // Clear the interrupt


I'm wondering if we can gain something if we store 1 << ISRlist[i] instead of ISRlist[i].

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/arduino/ArduinoCore-samd/pull/201#pullrequestreview-60574667>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AFfStIi7_bclt_6TnOHXgSrq-HZHwDmEks5sfTmNgaJpZM4Ld0iE>.

cmaglie commented:
Followups on #261
